[CIVP-11787] ENH futures.Executor classes for Civis

stephen-hoover opened the pull request:
Add two Civis-specific Executor classes (subclasses of the standard library `concurrent.futures.Executor`). The `_ContainerShellExecutor` uses Civis Platform container jobs to parallelize over either shell commands or container script arguments. The `CustomScriptExecutor` parallelizes over arguments to a custom script.

stephen-hoover commented:
@keithing , this is a lot of lines, but I think only the second commit needs a line-by-line scrutiny. Also, I plan to add documentation on the new features to the markdown pages in a future PR.

keithing commented on civis/futures.py:
```diff
@@ -285,3 +292,386 @@ def cancel(self):
             return False
+
+
+def create_docker_command(*args, **kwargs):
```
Do we want this to be private as well?

stephen-hoover replied:
You're right, it should be. Fixed.

keithing commented on civis/futures.py:
```diff
@@ -285,3 +292,386 @@ def cancel(self):
+        return job
+
+
+class CustomPoolExecutor(_CivisExecutor):
```
`CustomScriptPoolExecutor`?

stephen-hoover replied:
How about `CustomScriptExecutor`? It's shorter, and there's not really a "pool".

keithing replied:
Great, I like that.

keithing commented on civis/futures.py:
```diff
@@ -285,3 +292,386 @@ def cancel(self):
+
+        if script_name is None:
+            date_str = datetime.datetime.today().strftime("%Y-%m-%d")
+            script_name = "CustomPoolExecutorScript {}".format(date_str)
```
If we change the name, we need to update that here too.

keithing commented:
Everything looks good, after we decide what the correct name is for `CustomPoolExecutor` this looks ready to go.

stephen-hoover commented:
I renamed `CustomPoolExecutor` to `CustomScriptExecutor`. I also changed `_ContainerPoolShellExecutor` to `_ContainerShellExecutor`; it's shorter and probably more accurate.

keithing commented on civis/futures.py:
```diff
@@ -285,3 +292,386 @@ def cancel(self):
+
+        if script_name is None:
+            date_str = datetime.datetime.today().strftime("%Y-%m-%d")
+            script_name = ("ContainerPoolShellExecutorScript "
```
This needs to be renamed too.

stephen-hoover replied:
Oops. Fixed.
